toolkit/tools/internal/ccachemanager/ccachemanager.go

```diff
@@ -0,0 +1,769 @@
+				logger.Log.Infof("  (%s) is about to be de-referenced.", previousLatestTarSourcePath)
+			} else {
+				logger.Log.Warnf("  unable to get the latest ccache tag. This might be the first run and no latest ccache tag has been uploaded before.")
+			}
+		}
+
+		// Create the latest tag file...
+		logger.Log.Infof("  creating a tag file (%s) with content: (%s)...", m.CurrentPkgGroup.TagFile.LocalTargetPath, remoteStoreConfig.UploadFolder)
+		err = ioutil.WriteFile(m.CurrentPkgGroup.TagFile.LocalTargetPath, []byte(remoteStoreConfig.UploadFolder), 0644)
+		if err != nil {
+			return fmt.Errorf("Unable to write tag information to temporary file:\n%w", err)
+		}
+
+		// Upload the latest tag file...
+		logger.Log.Infof("  uploading tag file (%s) to (%s)...", m.CurrentPkgGroup.TagFile.LocalTargetPath, m.CurrentPkgGroup.TagFile.RemoteTargetPath)
+		err = m.AzureBlobStorage.Upload(context.Background(), m.CurrentPkgGroup.TagFile.LocalTargetPath, remoteStoreConfig.ContainerName, m.CurrentPkgGroup.TagFile.RemoteTargetPath)
+		if err != nil {
+			return fmt.Errorf("Unable to upload ccache archive:\n%w", err)
+		}
+
+		if remoteStoreConfig.KeepLatestOnly {
+			logger.Log.Infof("  keep latest only is enabled. Removing previous ccache archive if it exists...")
+			if previousLatestTarSourcePath == "" {
+				logger.Log.Infof("  cannot remove old archive with an empty name. No previous ccache archive to remove.")
+			} else {
+				logger.Log.Infof("  removing ccache archive (%s) from remote store...", previousLatestTarSourcePath)
+				err = m.AzureBlobStorage.Delete(context.Background(), remoteStoreConfig.ContainerName, previousLatestTarSourcePath)
+				if err != nil {
+					return fmt.Errorf("Unable to remove previous ccache archive:\n%w", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// After building a package or more, the ccache folder is expected to look as
+// follows:
+//
+// <rootDir> (i.e. /ccache)
+//
+//	<m.LocalDownloadsDir>
+//	<m.LocalUploadsDir>
+//	<m.RootWorkDir>
+//	  x86_64
+//	    <groupName-1>
+//	    <groupName-2>
+//	  noarch
+//	    <groupName-3>
+//	    <groupName-4>
+//
+// This function is typically called at the end of the build - after all
+// packages have completed building.
+//
+// At that point, there is not per package information about the group name
+// or the architecture.
+//
+// We use this directory structure to encode the per package group information
+// at build time, so we can use them now.
+func (m *CCacheManager) UploadMultiPkgGroupCCaches() (err error) {
+
+	architectures, err := directory.GetChildDirs(m.RootWorkDir)
+	errorsOccured := false
+	if err != nil {
+		return fmt.Errorf("failed to enumerate ccache child folders under (%s):\n%w", m.RootWorkDir, err)
+	}
+
+	for _, architecture := range architectures {
+		groupNames, err := directory.GetChildDirs(filepath.Join(m.RootWorkDir, architecture))
+		if err != nil {
+			logger.Log.Warnf("failed to enumerate child folders under (%s):\n%v", m.RootWorkDir, err)
```

Comment:
is `%v` intentional for printing err or should it be `%w`?

Reply:
same comment for a few other Warnf() calls in this file

Reply:
I get an error: `(*github.com/sirupsen/logrus.Logger).Warnf does not support error-wrapping directive %w`